src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -80,19 +82,153 @@ def from_setting(d: dict) -> "SRAMParameters":
             mux=int(d["mux"])
         )
 
+
 Supply = NamedTuple('Supply', [
     ('name', str),
     ('pin', Optional[str]),
     ('tie', Optional[str]),
     ('weight', Optional[int])
 ])
 
-PinAssignment = NamedTuple('PinAssignment', [
+
+class PinAssignmentError(ValueError):
+    """Exception raised from parsing an invalid PinAssignment object."""
+
+
+class PinAssignmentSemiAutoError(PinAssignmentError):
+    """Exception raised from parsing a PinAssignment that requires semi-auto features without explicitly enabing
+    them."""
+
+
+class PinAssignmentPreplacedError(PinAssignmentError):
+    """Exception raised from parsing a preplaced pin with extraneous information."""
+
+    def __init__(self, pin: "PinAssignment") -> None:
+        self.pin = pin
+
+    def __str__(self) -> str:
+        return "Pins {p} assigned as a preplaced pin with layers or side. Assuming pins are preplaced pins and ignoring layers and side.".format(
+            p=self.pin.pins)
+
+
+class PinAssignment(NamedTuple('PinAssignment', [
     ('pins', str),
     ('side', Optional[str]),
     ('layers', Optional[List[str]]),
-    ('preplaced', Optional[bool])
-])
+    ('preplaced', Optional[bool]),
+    ('location', Optional[Tuple[float, float]]),
+    ('width', Optional[float]),
+    ('depth', Optional[float])
+])):
+    __slots__ = ()
+
+    def __new__(cls, pins: str, side: Optional[str] = None, layers: Optional[List[str]] = None,
+                preplaced: Optional[bool] = None,
+                location: Optional[Tuple[float, float]] = None, width: Optional[float] = None,
+                depth: Optional[float] = None) -> "PinAssignment":
+        return super().__new__(cls, pins, side, layers, preplaced, location, width, depth)
+
+    @staticmethod
+    def create(pins: str, side: Optional[str] = None, layers: Optional[List[str]] = None,
+                preplaced: Optional[bool] = None,
+                location: Optional[Tuple[float, float]] = None, width: Optional[float] = None,
+                depth: Optional[float] = None) -> "PinAssignment":
+        """
+        Static method that works around the fact that mypy gets very confused at
+        the custom constructor above that defines default arguments.
```

Comment:
I'm still not sure I understand what's going on here- The comment is probably OK, but could you explain to me why mypy doesn't like the new constructor? Can you not make it happy with a type hint in `__new__` for some reason?

Reply:
I don't know the underlying technical reason, and I suspect it to be something internals to the workings of mypy. I think the upstream/desired fix is to move to Python3.6+ `NamedTuple`s or Python 3.7+ dataclasses, which we can't do (yet).